crates/sui-core/src/consensus_handler.rs

```diff
@@ -63,89 +65,104 @@ impl ConsensusHandler {
 impl ExecutionState for ConsensusHandler {
     /// This function will be called by Narwhal, after Narwhal sequenced this certificate.
     #[instrument(level = "trace", skip_all)]
-    async fn handle_consensus_transaction(
+    async fn handle_consensus_output(
         &self,
         // TODO [2533]: use this once integrating Narwhal reconfiguration
-        consensus_output: &Arc<ConsensusOutput>,
-        consensus_index: ExecutionIndices,
-        serialized_transaction: Vec<u8>,
+        consensus_output: ConsensusOutput,
     ) {
-        let index = Self::update_hash(
-            &self.last_seen,
-            consensus_index.clone(),
-            &serialized_transaction,
-        );
-        let index = if let Some(index) = index {
-            index
-        } else {
-            debug!(
+        let _scope = monitored_scope("HandleConsensusOutput");
+        let mut sequenced_transactions = Vec::new();
+        let mut seq = 0;
+
+        let round = consensus_output.sub_dag.round();
+        for (cert, batches) in consensus_output.batches {
+            let author = cert.header.author.clone();
+            let output_cert = Arc::new(cert);
+            for batch in batches {
+                for serialized_transaction in batch.transactions {
+                    let transaction = match bincode::deserialize::<ConsensusTransaction>(
+                        &serialized_transaction,
+                    ) {
+                        Ok(transaction) => transaction,
+                        Err(err) => {
+                            warn!(
+                                    "Ignoring malformed transaction (failed to deserialize) from {}: {}",
+                                    author, err
+                                );
+                            continue;
+                        }
+                    };
+                    let index = ExecutionIndices {
+                        last_committed_round: round,
+                        sub_dag_index: consensus_output.sub_dag.sub_dag_index,
+                        transaction_index: seq,
+                    };
+
+                    let index_with_hash =
+                        match Self::update_hash(&self.last_seen, index, &transaction.tracking_id) {
```

Comment:
I don't think it's a good idea to use tracking_id for integrity hash - there is no any guarantees about how other party calculates it. Let's just use serialized transaction content as before

Reply:
I see, ok